# ThreeSSeq-processing
This code repository contains scripts used to process ThreeSSeq data. Sequencing reads were first mapped to the E. coli K12 MG1655 genome (NCBI: NC_000913.3) using BWA-MEM (v0.7.12). PCR duplicates were removed by Picard Tool MarkDuplicates. Correlation analyses of sequencing data were performed using the functions multiBamSummary and plotCorrelation in deepTools(v3.0.1)17. For each individual ThreeSSeq experiment, reads were split based on strand using SAMtools. Top- and bottom-strand reads were processed separately. Peaks were called by MACS2 (v1.4.2) when adjusted P-value â‰¤ 0.00001, and those that overlap with non-specific peaks detected in the non-tagged DExI control were subtracted. Overlapping peak analyses were performed using the function findOverlaps in R package GenomicRanges.
